dask/dataframe/core.py

```diff
@@ -4340,26 +4405,44 @@ def repartition_npartitions(df, npartitions):
 
             return df.repartition(divisions=divisions)
         else:
-            ratio = npartitions / df.npartitions
-            split_name = 'split-%s' % tokenize(df, npartitions)
-            dsk = {}
-            last = 0
-            j = 0
-            for i in range(df.npartitions):
-                new = last + ratio
-                if i == df.npartitions - 1:
-                    k = npartitions - j
-                else:
-                    k = int(new - last)
-                dsk[(split_name, i)] = (split_evenly, (df._name, i), k)
-                for jj in range(k):
-                    dsk[(new_name, j)] = (getitem, (split_name, i), jj)
-                    j += 1
-                last = new
-
-            divisions = [None] * (npartitions + 1)
-            graph = HighLevelGraph.from_collections(new_name, dsk, dependencies=[df])
-            return new_dd_object(graph, new_name, df._meta, divisions)
+            div, mod = divmod(npartitions, df.npartitions)
+            nsplits = [div] * df.npartitions
+            nsplits[-1] += mod
+            return _split_partitions(df, nsplits, new_name)
+
+
+def _repartition_from_boundaries(df, new_partitions_boundaries, new_name):
+    if not isinstance(new_partitions_boundaries, list):
+        new_partitions_boundaries = list(new_partitions_boundaries)
+    if new_partitions_boundaries[0] > 0:
+        new_partitions_boundaries.insert(0, 0)
+    if new_partitions_boundaries[-1] < df.npartitions:
+        new_partitions_boundaries.append(df.npartitions)
+    dsk = {}
+    for i, (start, end) in enumerate(zip(new_partitions_boundaries, new_partitions_boundaries[1:])):
+        dsk[new_name, i] = (methods.concat, [(df._name, j) for j in range(start, end)])
+    divisions = [df.divisions[i] for i in new_partitions_boundaries]
+    graph = HighLevelGraph.from_collections(new_name, dsk, dependencies=[df])
+    return new_dd_object(graph, new_name, df._meta, divisions)
+
+
+def _split_partitions(df, nsplits, new_name):
+    if len(nsplits) != df.npartitions:
+        raise ValueError('nsplits should have len={}'.format(df.npartitions))
+
+    dsk = {}
+    split_name = 'split-{}'.format(tokenize(df, nsplits))
+    j = 0
+    for i, k in enumerate(nsplits):
+        # TODO: don't need to split for k==1
+        dsk[split_name, i] = (split_evenly, (df._name, i), k)
+        for jj in range(k):
+            dsk[new_name, j] = (getitem, (split_name, i), jj)
+            j += 1
+
+    divisions = [None] * (1 + sum(nsplits))
```

Comment:
It's somewhat unfortunate that we lose division information here.  I wonder if we might instead divide the range of a particular partition linearly, and then use those values to split up the partition rather than using something like `iloc`.  This can behave poorly if there is a large amount of skew within a partition, but I suspect that it will often be preferable.  

We could also leave this for future work if you prefer.

Reply:
Yes, I'd say let's leave it for now.